// ice/src/Glacier2/SSLInfo.js
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************
//
// Ice version 3.7.1
//
// <auto-generated>
//
// Generated from file `SSLInfo.ice'
//
// Warning: do not edit this file.
//
// </auto-generated>
//

/* eslint-disable */
/* jshint ignore: start */

/* slice2js browser-bundle-skip */
(function(module, require, exports)
{
/* slice2js browser-bundle-skip-end */
/* slice2js browser-bundle-skip */
    const _ModuleRegistry = require("../Ice/ModuleRegistry").Ice._ModuleRegistry;
    const Ice = _ModuleRegistry.require(module,
    [
        "../Ice/Struct",
        "../Ice/Long",
        "../Ice/HashMap",
        "../Ice/HashUtil",
        "../Ice/ArrayUtil",
        "../Ice/StreamHelpers",
        "../Ice/BuiltinSequences"
    ]).Ice;

    const Slice = Ice.Slice;
/* slice2js browser-bundle-skip-end */
/* slice2js browser-bundle-skip */

    let Glacier2 = _ModuleRegistry.module("Glacier2");
/* slice2js browser-bundle-skip-end */

    /**
     * Information taken from an SSL connection used for permissions
     * verification.
     *
     * @see PermissionsVerifier
     *
     **/
    Glacier2.SSLInfo = class
    {
        constructor(remoteHost = "", remotePort = 0, localHost = "", localPort = 0, cipher = "", certs = null)
        {
            this.remoteHost = remoteHost;
            this.remotePort = remotePort;
            this.localHost = localHost;
            this.localPort = localPort;
            this.cipher = cipher;
            this.certs = certs;
        }

        _write(ostr)
        {
            ostr.writeString(this.remoteHost);
            ostr.writeInt(this.remotePort);
            ostr.writeString(this.localHost);
            ostr.writeInt(this.localPort);
            ostr.writeString(this.cipher);
            Ice.StringSeqHelper.write(ostr, this.certs);
        }

        _read(istr)
        {
            this.remoteHost = istr.readString();
            this.remotePort = istr.readInt();
            this.localHost = istr.readString();
            this.localPort = istr.readInt();
            this.cipher = istr.readString();
            this.certs = Ice.StringSeqHelper.read(istr);
        }

        static get minWireSize()
        {
            return  12;
        }
    };

    Slice.defineStruct(Glacier2.SSLInfo, true, true);
/* slice2js browser-bundle-skip */
    exports.Glacier2 = Glacier2;
/* slice2js browser-bundle-skip-end */
/* slice2js browser-bundle-skip */
}
(typeof(global) !== "undefined" && typeof(global.process) !== "undefined" ? module : undefined,
 typeof(global) !== "undefined" && typeof(global.process) !== "undefined" ? require :
 (typeof WorkerGlobalScope !== "undefined" && self instanceof WorkerGlobalScope) ? self.Ice._require : window.Ice._require,
 typeof(global) !== "undefined" && typeof(global.process) !== "undefined" ? exports :
 (typeof WorkerGlobalScope !== "undefined" && self instanceof WorkerGlobalScope) ? self : window));
/* slice2js browser-bundle-skip-end */
